### PR TITLE
OTA-2540: IP Secondary integration tests

### PR DESCRIPTION
**Pull request by mike-sul**
Signed-off-by: Mike Sul <ext-mykhaylo.sul@here.com>

### CODE REVIEW

File: tests/ipsecondary_test.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._repo_root_dir.__exit__(exc_type, exc_val, exc_tb)
+
+    def _generate_repo(self):
+        subprocess.run([self._repo_manager_exe, '--path', self.root_dir, '--command', 'generate'], check=True)
+
+
+class UptaneTestBackend:
+
+    @property
```
Comment (lbonn) [inline tests/ipsecondary_test.py]: style question: what's the benefit of using this `@property` indirection here, compared to just defining `port` and `repo` members? Same question for constants like `image_rel_dir` of `UptaneTestRepo`.

Also, I've seen most __init__ of python classes declared at the top, is it also deliberate to have them in the middle of other methods?

File: tests/ipsecondary_test.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+            json.dump(sec_cfg, config_file)
+
+    def run(self, run_mode):
+        subprocess.run([self._aktualizr_primary_exe, '-c', self._config_file, '--run-mode', run_mode], check=True)
+
+    def get_info(self):
+        return str(subprocess.check_output([self._aktualizr_info_exe, '-c', self._config_file]))
+
+    # ugly stuff that could be removed if Aktualizr had exposed API to check status
+    # or aktializr-info had output status/info in a structured way (e.g. json)
```
Comment (patrickvacek) [inline tests/ipsecondary_test.py]: I would not at all be opposed to cleaning up the output of aktualizr-info. I would keep that type of thing out of aktualizr itself, though.

File: tests/ipsecondary_test.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+    # run aktualizr and secondary and wait until the device/aktualizr is registered
+    with aktualizr, secondary:
+        aktualizr.wait_for_completion()
+
+    # the secondary must be registered
+    if not aktualizr.is_ecu_registered(secondary.id):
+        return False
+
+    # run just aktualizr, the previously registered secondary is off
+    # and check if the primary ECU is updatable if the secondary is not connected
```
Comment (patrickvacek) [inline tests/ipsecondary_test.py]: How do you control/confirm that the secondary is not running here?

File: tests/ipsecondary_test.py (hunk 1 above)
Comment (mike-sul) [reply tests/ipsecondary_test.py]: > what's the benefit of using this @property indirection here, compared to just defining port and repo members?

make public API explicitly defined, so users/clients don't need to learn implementation in order to figure our a class interface/API

Comment (mike-sul) [reply tests/ipsecondary_test.py]: > Also, I've seen most init of python classes declared at the top, is it also deliberate to have them in the middle of other methods?

Yes, your're right, the standard approach/style is to put __init__ at the beginning of a class definition, I'll update my code correspondingly.

File: tests/ipsecondary_test.py (hunk 2 above)
Comment (mike-sul) [reply tests/ipsecondary_test.py]: ok. I'll elaborate my opinion during the stand-up meeting or via email.

File: tests/ipsecondary_test.py (hunk 3 above)
Comment (mike-sul) [reply tests/ipsecondary_test.py]: the context manager of the 'with' statement at line 486 guarantees that IPSecondary::__exit__ is called, but I'll add additional check to make sure that the secondary's process is not running.

File: tests/ipsecondary_test.py (hunk 1 above)
Comment (mike-sul) [reply tests/ipsecondary_test.py]: Just updated the PR, __init__ was moved up and useless properties are  removed

File: tests/ipsecondary_test.py (hunk 3 above)
Comment (mike-sul) [reply tests/ipsecondary_test.py]: Just updated

File: tests/ipsecondary_test.py (hunk 1 above)
Comment (lbonn) [reply tests/ipsecondary_test.py]: I don't want to insist on doing it at all cost but I also fail to see the changes right now. Github only shows a rebase without code changes when inspecting the last force-push.

Comment (mike-sul) [reply tests/ipsecondary_test.py]: Your comment is reasonable.
The code should be updated now, somehow I screwed up to commit the changes in the first place.

Comment (lbonn) [reply tests/ipsecondary_test.py]: Ok now I see it :+1:

**Comment by patrickvacek** [comment]
Ha! Just barely beat me to it.

**Pull request merged**